### PR TITLE
Add SAM.

### PR DESCRIPTION
**Pull request by Silver0930**

### CODE REVIEW

File: official/cv/segment-anything/README.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+## Installation
+
+The code requires `python>=3.7` and supports Ascend platform, some important pre-dependencies is: 
+1. mindspore: Please follow the instructions [here](https://www.mindspore.cn/install) to install mindspore dependencies.
+2. mindformers: please follow the instructions [here](https://gitee.com/mindspore/mindformers) using source code to install mindformers, 
+
+Clone the repository locally and install with
+
+```shell
+git clone https://github.com/Mark-ZhouWX/models.git
```
Comment (CaitinZhao) [inline official/cv/segment-anything/README.md:14]: 链接改掉

File: official/cv/segment-anything/README.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
+# Segment Anything
+
+The **Segment Anything Model (SAM)** produces high quality object masks from input prompts such as points or boxes, and it can be used to generate masks for all objects in an image. It has been trained on a [dataset](https://segment-anything.com/dataset/index.html) of 11 million images and 1.1 billion masks, and has strong zero-shot performance on a variety of segmentation tasks.
+
+## Installation
+
+The code requires `python>=3.7` and supports Ascend platform, some important pre-dependencies is: 
+1. mindspore: Please follow the instructions [here](https://www.mindspore.cn/install) to install mindspore dependencies.
```
Comment (CaitinZhao) [inline official/cv/segment-anything/README.md:8]: 版本信息用表格呈现：
| MindSPore | CANN | Mindformers |
| --- | --- | --- |
| [2.3.1](link) | [version]() | [version]() |

File: official/cv/segment-anything/README.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+### Finetune with box-prompt
+The bounding boxes are used as prompt input to predict mask.
+Beside fine-tuning our code on COCO2017 dataset which contains common seen objects and lies in the similar distribution of the original [training dataset](https://segment-anything.com/dataset/index.html) of SAM, We have done further experiments on a medical imaging segmentation dataset [FLARE22](https://flare22.grand-challenge.org/Dataset/). Result shows that the finetune method in this repository is effective.
+
+The bellowing shows the mask quality before and after finetune.
+
+
+| pretrained_model | dataset  |    epochs     | mIOU | ckpt                                                                                                         |
+|:----------------:| -------- |:-------------:|------|--------------------------------------------------------------------------------------------------------------|
+|    sam-vit-b     | COCO2017 | 0 (zero-shot) | 74.5 |                                                                                                              |
```
Comment (CaitinZhao) [inline official/cv/segment-anything/README.md:35]: zero shot的ckpt有吗

File: official/cv/segment-anything/README.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+A machine with **32G ascend memory** is required for box-prompt finetune.
+
+for standalone finetune of COCO dataset, please run:
+```shell
+python train.py -c configs/coco_box_finetune.yaml -o amp_level=O2
+```
+
+for distributed finetune of COCO dataset, please run:
+```shell
+msrun --worker_num=8 --local_worker_num=8 train.py -c configs/coco_box_finetune.yaml -o amp_level=O2
```
Comment (CaitinZhao) [inline official/cv/segment-anything/README.md:49]: 改成msrun的

File: official/cv/segment-anything/README.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+for standalone finetune of SA-1B dataset with BLIP2 (CLIP is similar), please run:
+```shell
+python train.py -c configs/sa1b_text_finetune_blip2.yaml
+```
+the BLIP2 checkpoint and bert vocabulary.txt will be automatically downloaded at `./checkpoint_download/`
+
+for distributed finetune, please run:
+```shell
+msrun --worker_num=8 --local_worker_num=8 train.py -c configs/sa1b_text_finetune_blip2.yaml
```
Comment (CaitinZhao) [inline official/cv/segment-anything/README.md:141]: msrun

File: official/cv/segment-anything/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+matplotlib
+pycocotools
+opencv-python
+omegaconf==2.0.0
+mindformers=1.0.0
```
Comment (CaitinZhao) [inline official/cv/segment-anything/requirements.txt:5]: 这个最好配套mindspore装

File: official/cv/segment-anything/README.md (hunk 4 above)
Comment (Silver0930) [reply official/cv/segment-anything/README.md:49]: 已经是msrun了呀

File: official/cv/segment-anything/requirements.txt (hunk 6 above)
Comment (Silver0930) [reply official/cv/segment-anything/requirements.txt:5]: done

Comment (Silver0930) [reply official/cv/segment-anything/requirements.txt:5]: > 这个最好配套mindspore装

done

File: official/cv/segment-anything/README.md (hunk 4 above)
Comment (Silver0930) [reply official/cv/segment-anything/README.md:49]: > 改成msrun的

已经是msrun了呀

File: official/cv/segment-anything/README.md (hunk 3 above)
Comment (Silver0930) [reply official/cv/segment-anything/README.md:35]: > zero shot的ckpt有吗

有，还没上传

File: official/cv/segment-anything/README.md (hunk 2 above)
Comment (Silver0930) [reply official/cv/segment-anything/README.md:8]: > 版本信息用表格呈现：
> 
> MindSPore	CANN	Mindformers
> [2.3.1](link)	version	version

done

File: official/cv/segment-anything/README.md (hunk 1 above)
Comment (Silver0930) [reply official/cv/segment-anything/README.md:14]: > 链接改掉

done

File: official/cv/segment-anything/README.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+### Using sam with Automatic Mask Generation(AMG)
+
+Since SAM can efficiently process prompts, masks for the entire image can be generated by sampling a large number of prompts over an image. AMG works by sampling single-point input prompts in a grid over the image, from each of which SAM can predict multiple masks. Then, masks are filtered for quality and deduplicated using non-maximal suppression. Additional options allow for further improvement of mask quality and quantity, such as running prediction on multiple crops of the image or postprocessing masks to remove small disconnected regions and holes.
+
+```shell
+python demo/inference_with_amg.py --model-type vit_h
+```
+
+See `python demo/inference_with_amg.py --help` to explore more custom settings.
```
Comment (CaitinZhao) [inline official/cv/segment-anything/README.md:247]: lite推理的代码有但是readme没有体现

File: official/cv/segment-anything/README.md [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+```shell
+msrun --worker_num=8 --local_worker_num=8 train.py -c configs/sa1b_text_finetune_blip2.yaml
+```
+the fine-tuned model will be saved at the work_root specified in `configs/sa1b_text_finetune.yaml`. For a fast single image inference, please run,
+
+```shell
+python text_inference.py --checkpoint=your/path/to/ckpt --text-prompt your_prompt
+```
+
+## Demo
```
Comment (CaitinZhao) [inline official/cv/segment-anything/README.md:151]: demo放finetune前面吧

File: official/cv/segment-anything/segment_anything/optim/group_param.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
+def create_group_param(params,  **kwargs):
```
Comment (SamitHuang) [inline official/cv/segment-anything/segment_anything/optim/group_param.py:1]: remove it if not use

File: official/cv/segment-anything/train.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+from segment_anything.utils import logger
+from segment_anything.utils.callbacks import create_callback
+from segment_anything.utils.config import parse_args
+from segment_anything.utils.model_wrapper import NetWithLossWrapper, TrainOneStepCellWrapper, SamIterativeSegModel
+from segment_anything.utils.utils import set_distributed, set_directory_and_log, update_rank_to_dataloader_config
+
+
+def main(args) -> None:
+    # Step1: initialize environment
+    ms.context.set_context(mode=args.mode, device_target=args.device, pynative_synchronize=True, max_device_memory="50GB")
```
Comment (SamitHuang) [inline official/cv/segment-anything/train.py:21]: why set pynative_synchronize=True? isn't it mainly for debug?
why limite max_device_memory to 50GB?

Comment (Silver0930) [reply official/cv/segment-anything/train.py:21]: > why set pynative_synchronize=True? isn't it mainly for debug? why limite max_device_memory to 50GB?

pynative_synchronize=True 是之前测过pynative没有删掉，这个我去掉 
 max_device_memory=50GB，是之前如果不限制会报OOM，我再测下去掉还会不会oom

File: official/cv/segment-anything/segment_anything/optim/group_param.py (hunk 9 above)
Comment (Silver0930) [reply official/cv/segment-anything/segment_anything/optim/group_param.py:1]: > remove it if not use
AdmW 类需要用到此方法

File: official/cv/segment-anything/README.md (hunk 3 above)
Comment (Silver0930) [reply official/cv/segment-anything/README.md]: > > zero shot的ckpt有吗
> 
> 有，还没上传
已更新

File: official/cv/segment-anything/README.md (hunk 8 above)
Comment (Silver0930) [reply official/cv/segment-anything/README.md]: > demo放finetune前面吧

done

File: official/cv/segment-anything/train.py (hunk 10 above)
Comment (Silver0930) [reply official/cv/segment-anything/train.py]: > why set pynative_synchronize=True? isn't it mainly for debug? why limite max_device_memory to 50GB?

已更新

File: official/cv/segment-anything/README.md [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+python text_inference.py --checkpoint=your/path/to/ckpt --text-prompt your_prompt
+```
+
+### Performance
+
+Experiments are tested on ascend910* with mindspore 2.3.1 graph mode.
+
+| model name | cards | batch size | graph compile | recompute | sink | jit level | s/step | img/s |
+|:----------:| :---: | :--------: | :-----------: | :-------: | :---: | :------: | :----: | :---: |
+| sam-vit-b finetune on COCO2017 | 1 | 16 | 3~4 mins | OFF | ON | O2 | 2.59 | 6.18 |
```
Comment (SamitHuang) [inline official/cv/segment-anything/README.md:253]: better add a `dataset` column instead of writing the dataset name in `model_name` column

File: official/cv/segment-anything/README_INFERENCE.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
+# Inference
+
+## Installation
+
+The code supports Ascend platform, some important pre-dependencies is:
```
Comment (SamitHuang) [inline official/cv/segment-anything/README_INFERENCE.md:5]: same here. fix grammar

File: official/cv/segment-anything/README_INFERENCE.md [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+| mindspore-lite | mindspore | ascend driver | firmware | cann tookit/kernel |
+| :------------: | :-------: | :-----------: | :------: | :----------------: |
+| 2.3.1 | 2.3.1 | Same as Finetune | Same as Finetune | Same as Finetune |
```
Comment (SamitHuang) [inline official/cv/segment-anything/README_INFERENCE.md:9]: Same as Finetune? 
use hyperlink or simply specify the version code

File: official/cv/segment-anything/README_INFERENCE.md [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+See `python lite_inference.py --help` to explore more custom settings.
+
+## Performance
+
+Experiments are tested on ascend310* with mindspore-lite 2.3.1 .
+
+| model name | s/img |
+|:----------:| :---: |
+| sam-vit-b finetune on COCO2017 | 5.35 |
```
Comment (SamitHuang) [inline official/cv/segment-anything/README_INFERENCE.md:47]: same here.
or finetuned on xxx

File: official/cv/segment-anything/demo/inference_with_prompts.py [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+    model_type = args.model_type
+    sam_checkpoint = args.checkpoint
+
+    print(f'running with {model_type}')
+
+    sam = sam_model_registry[model_type](checkpoint=sam_checkpoint)
+
+    predictor = SamPredictor(sam)
+
+    image = cv2.imread('https://github.com/user-attachments/assets/022bd54a-4e28-4c22-82d4-45113bbe49ee')
```
Comment (SamitHuang) [inline official/cv/segment-anything/demo/inference_with_prompts.py:27]: Allow setting input image via argparser, same as lite inference `--image_path`

File: official/cv/segment-anything/point_inference.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+        image_path = args.image_path
+        image_np = cv2.imread(image_path) # (1024, 1024, 3)
+        image_np = cv2.cvtColor(image_np, cv2.COLOR_BGR2RGB)
+
+        # image = np.load(image_path)
+        # cv2.imwrite('./images/tumor1.png', image_np*255)
+        # image_npnew = cv2.imread('./images/tumor1.png')/255.0  # (1024, 1024, 3)
+        # diff = image_npnew - image_np
+
+        points_np = np.array([[[868, 622]], [[724, 466]], [[501, 543]], [[565, 505]], [[262, 456]]]) #  (bp, n, 2), in [w, h]
```
Comment (SamitHuang) [inline official/cv/segment-anything/point_inference.py:37]: is this point prompt bound with `images/tumor1.png`? 
what if user change `image_path` to another image?
better allow setting point prompt via argparser too.

**Review by SamitHuang — COMMENTED** [review]
minor reject

**Review by vigo999 — APPROVED** [review]
